Sections 8.7 - 8.10: synchronization, ordering, locking

spotluri opened the pull request:
fixing mislabled PEs in quiet.pdf - addressing github issue #81

min-si commented:
@spotluri : Thanks for creating the PR. I've created issue #127  for the fortran description issue.

min-si commented:
@spotluri : Looks I do not have permission to add label or project card. Could you please mark PR [#128](https://github.com/openshmem-org/specification/pull/128) for me ? Thanks.

min-si commented:
@spotluri : Here is another minor fix PR [#129](https://github.com/openshmem-org/specification/pull/129). Could you please label/card it as well ?

spotluri commented:
@min-si: i have added the cards and also merged them into the section PR. please review.

min-si commented:
Thanks !

min-si commented:
When I was updating the nonblocking Put/Get in the memory ordering section, I was a little confused whether `shmem_fence` ensures ordering of nonblocking Get operations. It is also not explained in the RMA section.
Specifically, does fence ensure ordering in any of the follow cases ? 
1. Gets issued to the same remote pe:
```
shmem_get_nbi(loc_a, rem_a,...,pe0);
shmem_fence();
shmem_get_nbi(loc_b, rem_b,...,pe0);
```
2. Gets issued to different remote pe:
```
shmem_get_nbi(loc_a, rem_a,...,pe0);
shmem_fence();
shmem_get_nbi(loc_b, rem_b,...,pe1);
```
3. Get and Put issued to the same pe:
```
shmem_get_nbi(loc_a, rem_a,...,pe0);
shmem_fence();
shmem_put_nbi(rem_b, loc_b,...,pe0);
```

spotluri commented:
@min-si sorry for pitching in late. 

My understanding of the current spec is that fence does not order get operations. So, it does not guarantee  ordering in any of the three cases you mention. 

@shamisp / @manjugv do you remember if this gets were intentionally left out of fence ordering. 

A change to this will be a semantic change and we should create a regular ticket if we would like to make this change.

spotluri commented:
Hi All, 

Attached is a PDF of sections on synchronization, ordering, locking, now 8.8-8.11. I have merged all PRs and addressed issues raised until now. I have rebased it to the ToT of osh_spec_next. 

[section-sync-review-1.4.pdf](https://github.com/openshmem-org/specification/files/1283061/section-sync-review-1.4.pdf)

@min-si, @spophale, @shamisp, @manjugv, @abouteiller, can you please give your approval (thumbs up) on the draft. The deadline is Sept 7.

manjugv commented:
> @min-si sorry for pitching in late.
> 
> My understanding of the current spec is that fence does not order non-blocking get operations. So, it does not guarantee ordering in any of the three cases you mention.
> 
> @shamisp / @manjugv do you remember if gets were intentionally left out of fence ordering.
> 
@min-si @spotluri That is correct. The fence does not order the nonblocking `Get` operations. This is to keep consistent with the blocking `Get` semantics.

manjugv commented:
@spotluri You might have to remove Get from the commit, otherwise 👍

jdinan commented:
How about adding a sentence to clarify that nonblocking gets are not ordered by fence?

spotluri commented:
@manjugv, the draft does not add non-blocking gets to the list of operations that shmem_fence orders. non-blocking get have only been added to quiet. Am I missing an instance?

spotluri commented:
objects to the same \ac{PE}. \FUNC{shmem\_fence} guarantees order of delivery,
-    not completion.
+    not completion. It does not guarantee order of delivery of non-blocking \GET{} routines.

manjugv commented:
@spotluri I was referring to introduction paragraph main_spec.tex line 288/289. But, I guess that is OK.

abouteiller commented:
Could we move table 5 and 6 above 8.9.1?

88:6: by a transfer from a remote PE -> by a PE
  rationale: could be a thread of the same PE

90:30: the following example uses shmem_fence in a C11 program.

Shouldn't the block of text in 8.10.2 page 91:35-40 also appear in shmem_fence ? (text about ordering of memory stores between multiple threads wrt shmem_quiet). 

92:30: "by the program user" -> in the user's program

spotluri commented:
@abouteiller, @jdinan: thanks for the feedback. 

@abouteiller: I have merged your changes into the section PR. 

@jdinan: I have added a sentence clarifying that non-blocking gets are not ordered by fence.

@min-si, @spophale, @shamisp, @manjugv, @abouteiller, here is an updated draft for your review and approval

[section-sync-review-1.4-v2.pdf](https://github.com/openshmem-org/specification/files/1286209/section-sync-review-1.4-v2.pdf)

shamisp commented:
👍

manjugv commented:
👍

min-si commented:
Sorry for my late response. I was on vacation in this week. The draft looks great to me.

spotluri commented on content/shmem_fence.tex:
```diff
@@ -1,5 +1,5 @@
 \apisummary{
-    Assures ordering of delivery of \PUT{}, \acp{AMO}, and memory store routines
+    Assures ordering of delivery of \PUT{}, \ac{AMO}, memory store, and non-blocking \PUT{} routines
     to symmetric data objects.
```
non-blocking -> nonblocking

spotluri replied:
fixed

jdinan commented on content/shmem_fence.tex:
```diff
@@ -1,5 +1,5 @@
 \apisummary{
-    Assures ordering of delivery of \PUT{}, \acp{AMO}, and memory store routines
+    Assures ordering of delivery of \PUT{}, \ac{AMO}, memory store, and non-blocking \PUT{} routines
```
Update "non-blocking" to "nonblocking"

nspark commented on content/shmem_fence.tex:
```diff
@@ -43,19 +43,29 @@
 
 \apicexample
-    {The following \FUNC{shmem\_fence} example is for C11 programs: }
+    {The following example uses  \FUNC{shmem\_fence}  in a C11 program: }
```
`C11` --> `\Cstd[11]`

nspark replied:
Similarly, this change should apply to `shmem_quiet.tex`.

BryantLam commented on content/shmem_fence.tex:
```diff
@@ -43,19 +43,29 @@
 
 \apicexample
-    {The following \FUNC{shmem\_fence} example is for C11 programs: }
+    {The following example uses  \FUNC{shmem\_fence}  in a C11 program: }
```
\Cstd[11]

abouteiller commented on content/shmem_fence.tex:
```diff
@@ -43,19 +43,29 @@
+    \ac{AMO}, memory store, and non-blocking \PUT{} routines to symmetric data objects is desired
     when multiple remote \acp{PE} are involved.
+
+    In an \openshmem program with multithreaded \acp{PE}, it is the
```
@manjugv Did you have time to review this piece of text?

spotluri replied:
i have closed the issue based on Manju's feedback

nspark commented:
@spotluri I've opened https://github.com/spotluri/openshmem-specification/pull/11 to fix #175 and also noticed that you missed the editorial changes in https://github.com/spotluri/openshmem-specification/pull/9 that I opened after RCM1.

spotluri commented:
@nspark: sorry, I did miss the post-RC1 changes PR. Merged it in now. 

 spotluri#11 looks good to me. Will wait on an approval from @jamesaross before merging it in.
